compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -403,21 +403,37 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
       else
         Prod(pat.tpe.stripAnnots, fun.tpe.widen, fun.symbol, pats.map(project), irrefutable(fun))
     case Typed(pat @ UnApply(_, _, _), _) => project(pat)
-    case Typed(expr, _) => Typ(erase(expr.tpe.stripAnnots), true)
+    case Typed(expr, tpt) =>
+      val unchecked = expr.tpe.hasAnnotation(ctx.definitions.UncheckedAnnot)
+      def warn(msg: String): Unit = if (!unchecked) ctx.warning(UncheckedTypePattern(msg), tpt.pos)
+      Typ(erase(expr.tpe.stripAnnots)(warn), true)
     case _ =>
       debug.println(s"unknown pattern: $pat")
       Empty
   }
 
   /* Erase a type binding according to erasure semantics in pattern matching */
-  def erase(tp: Type): Type = tp match {
+  def erase(tp: Type)(implicit warn: String => Unit): Type = tp match {
```

Comment:
Instead of passing an implicit around, you could have an outer function which defines `warn` and an inner recursive function which uses it.

Reply:
@smarter thanks a lot, I've addressed other comments.  For this one, doing the refactoring would require to pass an additional `Position` to `erase`, which will make the API a little ugly.